core/arch/clean_call_opt_shared.c

```diff
@@ -732,14 +732,11 @@ analyze_clean_call(dcontext_t *dcontext, clean_call_info_t *cci, instr_t *where,
     }
 # ifdef X86
     /* 9. derived fields */
-    /* XXX: For x64, skipping a single reg or flags still results in a huge
-     * code sequence to put in place: we may want to still use out-of-line
-     * unless multiple regs are able to be skipped.
+    /* Use out-of-line calls if more than 3 SIMD or GP registers have to be saved.
      * XXX: This should probably be in arch-specific clean_call_opt.c.
      */
-    if ((cci->num_simd_skip == 0 /* save all xmms */ &&
-         cci->num_regs_skip == 0 /* save all regs */ &&
-         !cci->skip_save_flags) ||
+    if ((NUM_SIMD_REGS - cci->num_simd_skip) > 3  /* save more than 3 xmms */ ||
```

Comment:
On x86_32, a SIMD save/restore is 6-9 bytes, a GPR save/restore is 3-6 bytes (but we could use pusha to save all regs), aflags save/restore is 1 byte (pushf/popf). 
On x86_64, a SIMD save/restore is 6-9 bytes, a GPR save/restore is 4-7 bytes, aflags 1 byte.
A call instruction is 5 byte.
So just for the size, we could have a really low bar.
For performance, we might want to a bit higher bar, assuming a full context switch is performed for out-of-line swap.
So current number is not unreasonable.
sum is not a bad idea, we might want to use weighted sum with the number of bytes for each reg save/restore, and the threshold could be 25 bytes or something.